src/main/java/com/zaxxer/nuprocess/linux/EpollEvent.java

```diff
@@ -19,54 +19,66 @@
 import java.util.Arrays;
 import java.util.List;
 
-import com.sun.jna.Pointer;
 import com.sun.jna.Structure;
 import com.sun.jna.Union;
 
 public class EpollEvent extends Structure
 {
 
    /*
-       typedef union epoll_data
-       {
-         void *ptr;
-         int fd;
-         uint32_t u32;
-         uint64_t u64;
-       } epoll_data_t;
-    
        struct epoll_event
        {
          uint32_t events;   // Epoll events
          epoll_data_t data; // User data variable
-       };
+       } __EPOLL_PACKED;
+
+       On x86, __EPOLL_PACKED is:    #define __EPOLL_PACKED
+       On x86_64, __EPOLL_PACKED is: #define __EPOLL_PACKED __attribute__ ((__packed__))
+
+       sizeof(struct epoll_event) is 12 on x86 and x86_64
    */
 
    public int events;
    public EpollData data;
 
    EpollEvent()
    {
-      // per eventpoll.h, x86_64 has the same alignment as 32-bit
-      super(ALIGN_GNUC);
+      // per bits/epoll.h, epoll_event is created with __attribute__ ((__packed__)), which disables
+      // applying padding to optimize alignment. epoll_event is memory-aligned on 32-bit platforms,
+      // but not on 64-bit platforms (i.e. it uses 32-bit alignment on 64-bit platforms)
+      super(ALIGN_NONE);
```

Comment:
I'm still doing some testing on 32-bit Linux to further verify this change

Reply:
Let us know when you've finished 32-bit verification.

Reply:
Sorry, took a little bit of setup (I didn't have a 32-bit VM on hand) but I've just finished up testing it now. All of the unit tests pass on x86 Ubuntu 14.04. Once I've added the benchmark I wrote to the project I'll run that on my 32-bit environment as well for a little more stress testing.

Details for my 32-bit environment:

```
bturner@ubuntu:~/Development/oss/nuprocess$ uname -a
Linux ubuntu 4.4.0-31-generic #50~14.04.1-Ubuntu SMP Wed Jul 13 01:06:37 UTC 2016 i686 athlon i686 GNU/Linux
bturner@ubuntu:~/Development/oss/nuprocess$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/i686-linux-gnu/4.8/lto-wrapper
Target: i686-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.8.4-2ubuntu1~14.04.3' --with-bugurl=file:///usr/share/doc/gcc-4.8/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.8 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.8 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libmudflap --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.8-i386/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.8-i386 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.8-i386 --with-arch-directory=i386 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-targets=all --enable-multiarch --disable-werror --with-arch-32=i686 --with-multilib-list=m32,m64,mx32 --with-tune=generic --enable-checking=release --build=i686-linux-gnu --host=i686-linux-gnu --target=i686-linux-gnu
Thread model: posix
gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04.3)
```

A slightly modified version of my `sizeof` program from #79 confirms (as expected) that the `sizeof(struct epoll_event)` is 12 on x86 Linux as well.

Reply:
I've run my jmh benchmark on my 32-bit VM now. Here's the summary:

```
# Run complete. Total time: 00:04:07

Benchmark                                                 Mode  Cnt       Score        Error    Units
NuProcessBenchmark.cat                                   thrpt  100      20.496 ±      0.486    ops/s
NuProcessBenchmark.cat:·gc.alloc.rate                    thrpt  100       7.703 ±      0.183   MB/sec
NuProcessBenchmark.cat:·gc.alloc.rate.norm               thrpt  100  540022.366 ±    472.536     B/op
NuProcessBenchmark.cat:·gc.churn.Metaspace               thrpt  100      ≈ 10⁻⁴                MB/sec
NuProcessBenchmark.cat:·gc.churn.Metaspace.norm          thrpt  100       2.786 ±      6.647     B/op
NuProcessBenchmark.cat:·gc.churn.PS_Eden_Space           thrpt  100       6.662 ±      4.691   MB/sec
NuProcessBenchmark.cat:·gc.churn.PS_Eden_Space.norm      thrpt  100  518120.933 ± 367662.475     B/op
NuProcessBenchmark.cat:·gc.churn.PS_Old_Gen              thrpt  100       0.520 ±      0.884   MB/sec
NuProcessBenchmark.cat:·gc.churn.PS_Old_Gen.norm         thrpt  100   40171.064 ±  67837.749     B/op
NuProcessBenchmark.cat:·gc.churn.PS_Survivor_Space       thrpt  100       0.057 ±      0.075   MB/sec
NuProcessBenchmark.cat:·gc.churn.PS_Survivor_Space.norm  thrpt  100    4530.189 ±   5891.110     B/op
NuProcessBenchmark.cat:·gc.count                         thrpt  100      29.000                counts
NuProcessBenchmark.cat:·gc.time                          thrpt  100    4980.000                    ms
NuProcessBenchmark.cat:·threads.alive                    thrpt  100      18.460 ±      0.170  threads
NuProcessBenchmark.cat:·threads.daemon                   thrpt  100      17.460 ±      0.170  threads
NuProcessBenchmark.cat:·threads.started                  thrpt  100      19.000               threads
```

20 ops/sec, given each op is a complete `cat` of a 250MB test file, translates to a throughput of roughly 5GB/s (using the default 4 pump threads, given my VM has 8 VCPUs). Not bad

Reply:
Not bad.  If you need higher throughput, you can experiment with calling fcntl with F_SETPIPE_SZ on the pipe.  The Linux default pipe size is 64K, but can be set from user-space (via fcntl) up to 1MB.  There’s probably a sweet spot in there, maybe between 128-256K.

Reply:
Of course you would need to adjust the NuProcess buffer sizes to match.

Reply:
Actually, thinking about it, something like “cat” pretty much produces constant output, but a git command’s output is likely to be quite “bursty”.  A burst of output larger than 64K will stall the git process until NuProcess can drain the pipe.  So setting a larger pipe size might have good payoff there.

Reply:
Git's an interesting case, because it doesn't behave quite how one might expect. For hosting operations (which, for obvious reasons, is what I care about the most here) it tends to reach a "steady state" fairly quickly. After the ref advertisement, it generates a pack with the requested objects and from there it's just streaming that pack. The wire protocol uses a packet format, which is bounded to 8K. That means it writes _at most_ a shade over 8K (8197 bytes) at a time.

We support two different transports, HTTP (Tomcat) and SSH (Mina SSHD). Tomcat uses an 8K buffer by default for `HttpServletResponse`, so we use an 8K buffer when copying streams. For SSH, specifically because of Git's packet sizes, we're using an exact 8197 byte buffer to copy streams. These buffers effectively carry any backpressure from the socket through to the underlying Git process. If the socket can't move the data fast enough, the Git process gets blocked trying to write more output and slows down to match.

Of course, as the pump threads fall behind (which they almost certainly would, since they'd be marshaling over to a socket of some sort), while Git might not be writing more than ~8K per shot, that doesn't mean there can't be multiple packets waiting for the pump thread to read. But since those larger reads are still going to be funneled through a buffer somewhere around 8K on their way to the client, those larger reads may not offer much benefit (and, in fact, doing smaller, more packet-sized reads might allow the pump thread to dump output to more sockets per second, keeping perceived transfer rates more steady).

Setting hosting aside, a lot of what Bitbucket Server spends the rest of its time doing is running `git rev-list`, `git  diff-tree` and `git diff`, with a few other commands interwoven here and there. For those commands, it's entirely possible they won't produce 64K of output over their _entire execution_, let alone in a single shot.

If anything, then, I think what we'd really want is a way to change `NuProcess.BUFFER_CAPACITY` from a hard constant to something that can, at least to some extent, be changed at runtime. The most flexible option would probably be to make it configurable on the `NuProcessBuilder` and use `NuProcess.BUFFER_CAPACITY` as a default, but even making it configurable via system property would let the application do some tuning.